### PR TITLE
Fix/small fixes

### PR DESCRIPTION
**Pull request by ad2969**
Addresses some of the small issues in #102:

- Upgrades `formik` package to newer version (apparently solves some `async` detecting issues on submitting)
- Uses `formik` props to disable button when submitting/no fields are changed

- Added a loading buffer on the router
(when the page is refreshed, for a split second the login page is shown - the "loading" buffer adds a state that tracks whether all the data is loaded)

- Abstracted the loading circle anim to its own small component

### CODE REVIEW

File: src/components/Router.js [hunk 1]
```diff
@@ -45,24 +52,41 @@ class Router extends Component {
       .catch(() => log('Not signed in'))
   }
 
-  componentDidMount() {
-   getEvents()
+  // User needs to be checked before the page physically renders
+  // (otherwise, the login page will initially show on every refresh)
+  componentWillMount() {
```
Comment (ianmah) [inline src/components/Router.js:57]: Does this need to be `willMount`? I think the `loaded` state will block rendering a login page regardless, and `willMount` is technically [deprecated](https://reactjs.org/docs/react-component.html#legacy-lifecycle-methods)

Comment (ad2969) [reply src/components/Router.js]: True true

File: src/components/Loading/index.js [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+  }));
+
+export default function Loading(props) {
+
+    const classes = useStyles();
+
+    return (
+        <div className={classes.layout}>
+            <CircularProgress />
+            <span className={classes.textContent}>{props.message}</span>
```
Comment (ianmah) [inline src/components/Loading/index.js:28]: Can we use `Typography` component so the styles will be more consistent?

File: src/components/Router.js [hunk 3]
```diff
@@ -45,24 +53,41 @@ class Router extends Component {
...
+  // User needs to be checked before the page physically renders
+  // (otherwise, the login page will initially show on every refresh)
   componentDidMount() {
-   getEvents()
 
-   if(!this.props.user) this.getAuthenticatedUser();
+   if(!this.props.user) {
+    // If the user doesn't already exist in react, get the authenticated user
+     this.getAuthenticatedUser()
+      // After getting the authenticated user, get the events
```
Comment (ianmah) [inline src/components/Router.js:63]: Can we get the user and event at the same time?

File: src/components/Router.js [hunk 4]
```diff
@@ -45,24 +53,41 @@ class Router extends Component {
...
+      .then(getEvents)
+      // Ultimately, after all is loaded, set the "loaded" state and render the component
+      .then(() => {
```
Comment (ianmah) [inline src/components/Router.js:66]: could use a Promise.all if we can get user and event at the same time

File: src/components/Authentication/LoginRedirect.js [hunk 5]
```diff
@@ -1,105 +1,111 @@
...
 import { Auth } from "aws-amplify";
 import { setUser } from '../../actions/UserActions'
 import { connect } from "react-redux";
+import Loading from "../Loading";
 import { withRouter } from "react-router-dom";
-import CircularProgress from '@material-ui/core/CircularProgress';
 import { log, fetchBackend } from '../../utils'
 
-export class LoginRedirect extends Component {
+function LoginRedirect(props) {
```
Comment (ad2969) [inline src/components/Authentication/LoginRedirect.js:9]: Basically turning this into a functional comp instead

File: src/components/Authentication/LoginRedirect.js [hunk 6]
```diff
@@ -1,105 +1,111 @@
...
-                                    console.log(err.status)
+                                    console.log('Encountered an error querying database!', err.status);
                                 }
-                            })
-                    } else {
-                        // "First time sign up" aka no student_id in user pool
-                        clearTimeout(this.timeoutRedirect)
-                        this.populateUserAndRedirect(authUser)
+                            }
+                        } else {
```
Comment (ad2969) [inline src/components/Authentication/LoginRedirect.js:79]: not sure if we want to cover the end case when the user "doesnt exist in the user pool, but exists in database"

File: src/utils/index.js [hunk 7]
```diff
@@ -57,25 +57,24 @@ export function log(message) {
 }
 
 // Refresh the redux store
-export function updateEvents() {
-    fetchBackend('/events', 'GET')
-        .then(response => {
+export async function updateEvents() {
```
Comment (ianmah) [inline src/utils/index.js:60]: does this return a promise? Wondering if wrapping this inside a Promise.all is actually correct

File: src/components/Authentication/LoginRedirect.js (hunk 6 above)
Comment (ianmah) [reply src/components/Authentication/LoginRedirect.js:79]: nah

File: src/utils/index.js (hunk 7 above)
Comment (ad2969) [reply src/utils/index.js:60]: this confused me for the longest time LOL.. calling an async function without using `await` returns a promise

Comment (ianmah) [reply src/utils/index.js:60]: And we want to return a promise right?

Comment (ad2969) [reply src/utils/index.js:60]: yup.. `Promise.all` accepts an iterable of promises

**Pull request merged**